subiquity/common/filesystem/tests/test_manipulator.py

```diff
@@ -393,32 +420,93 @@ def DONT_test_add_boot_BIOS_preserved(self):  # needs v2 partitioning
+    def test_add_boot_BIOS_new_and_preserved_v2(self):
+        manipulator = make_manipulator(Bootloader.BIOS, 2)
+        # 2002MiB so that the space available for partitioning (2000MiB)
+        # divided by 4 is an whole number of megabytes.
+        disk = make_disk(manipulator.model, preserve=True, size=2002*MiB)
+        avail = gaps.largest_gap_size(disk)
+        p1_new = make_partition(
+            manipulator.model, disk, size=avail//4)
+        p2_preserved = make_partition(
+            manipulator.model, disk, size=3*avail//4, preserve=True)
         with self.assertPartitionOperations(
                 disk,
                 Create(
                     offset=disk.alignment_data().min_start_offset,
-                    size=size),
+                    size=sizes.BIOS_GRUB_SIZE_BYTES),
+                MoveResize(
+                    part=p1_new,
+                    offset=sizes.BIOS_GRUB_SIZE_BYTES,
+                    size=-sizes.BIOS_GRUB_SIZE_BYTES),
+                Unchanged(part=p2_preserved),
                 ):
             manipulator.add_boot_disk(disk)
         self.assertIsBootDisk(manipulator, disk)
 
-    def test_add_boot_UEFI_empty(self):
-        manipulator = make_manipulator(Bootloader.UEFI)
+    @parameterized.expand([(1,), (2,)])
+    def test_no_add_boot_BIOS_preserved_at_start(self, version):
+        manipulator = make_manipulator(Bootloader.BIOS, version)
         disk = make_disk(manipulator.model, preserve=True)
-        self._test_add_boot_empty(
-            manipulator, disk, sizes.get_efi_size(disk))
+        half_size = gaps.largest_gap_size(disk)//2
+        make_partition(
+            manipulator.model, disk, size=half_size,
+            preserve=True)
+        self.assertFalse(boot.can_be_boot_device(disk))
+
+    @parameterized.expand([(1,), (2,)])
+    def test_add_boot_BIOS_msdos(self, version):
+        manipulator = make_manipulator(Bootloader.BIOS, version)
+        disk = make_disk(manipulator.model, preserve=True, ptable='msdos')
+        part = make_partition(
+            manipulator.model, disk, size=gaps.largest_gap_size(disk),
+            preserve=True)
+        with self.assertPartitionOperations(
+                disk,
+                Unchanged(part=part),
+                ):
+            manipulator.add_boot_disk(disk)
+        self.assertIsBootDisk(manipulator, disk)
 
-    def test_add_boot_PREP_empty(self):
-        manipulator = make_manipulator(Bootloader.PREP)
+    def boot_size_for_disk(self, disk):
+        bl = disk._m.bootloader
+        if bl == Bootloader.UEFI:
+            return sizes.get_efi_size(disk)
+        elif bl == Bootloader.PREP:
+            return sizes.PREP_GRUB_SIZE_BYTES
+        else:
+            self.fail("unexpected bootloader %r" % (bl,))
+
+    @parameterized.expand([
+        (Bootloader.UEFI, 1),
+        (Bootloader.PREP, 2),
+        (Bootloader.UEFI, 1),
+        (Bootloader.PREP, 2)])
```

Comment:
Was this intended as below?
See also 6 repeated copies of this table later in the code
```suggestion
        (Bootloader.UEFI, 1),
        (Bootloader.PREP, 1),
        (Bootloader.UEFI, 2),
        (Bootloader.PREP, 2)])
```

Reply:
Ah haha oops.